tensorboard/plugins/hparams/download_data_test.py

```diff
@@ -0,0 +1,295 @@
+    }
+    metric_values { name { tag: "delta_temp" } value: -151000000.0
+      training_step: 2
+      wall_time_secs: 10.0
+    }
+  }
+}
+total_size: 3
+"""
+
+
+EXPECTED_LATEX = r"""\begin{table}[tbp]
+\begin{tabular}{llllllll}
+initial\_temp & final\_temp & string\_hparam & bool\_hparam & optional\_string\_hparam & current\_temp & delta\_temp & optional\_metric \\ \hline
+$270$ & $150$ & a string & $1$ &  & $10$ & $15$ & $33$ \\
+$280$ & $100$ & AAAAA & $0$ &  & $51$ & $44.5$ \\
+$300$ & $1.2\cdot 10^{-5}$ & a string\_3 & $1$ & BB & $101$ & $-1.51\cdot 10^{7}$ \\
+\hline
+\end{tabular}
+\end{table}
+"""
+
+EXPECTED_CSV = """initial_temp,final_temp,string_hparam,bool_hparam,optional_string_hparam,current_temp,delta_temp,optional_metric\r
+270.0,150.0,a string,True,,10.0,15.0,33.0\r
+280.0,100.0,AAAAA,False,,51.0,44.5\r
+300.0,1.2e-05,a string_3,True,BB,101.0,-15100000.0\r
+"""
+
+
+class DownloadDataTest(tf.test.TestCase):
+    def setUp(self):
+        self._mock_tb_context = mock.create_autospec(base_plugin.TBContext)
+        self._mock_multiplexer = mock.create_autospec(
+            plugin_event_multiplexer.EventMultiplexer
+        )
+        self._mock_tb_context.multiplexer = self._mock_multiplexer
+
+    def _run_handler(self, experiment, session_groups, response_format):
+        experiment_proto = text_format.Merge(experiment, api_pb2.Experiment())
+        session_groups_proto = text_format.Merge(
+            session_groups, api_pb2.ListSessionGroupsResponse()
+        )
+        handler = download_data.Handler(
+            backend_context.Context(self._mock_tb_context),
+            experiment_proto,
+            session_groups_proto,
+            response_format,
+        )
+        return handler.run()
+
+    def test_csv(self):
+        body, mime_type = self._run_handler(
+            EXPERIMENT, SESSION_GROUPS, download_data.OutputFormat.CSV
+        )
+        self.assertEqual("text/csv", mime_type)
+        self.assertEqual(EXPECTED_CSV, body)
+
+    def test_latex(self):
+        body, mime_type = self._run_handler(
+            EXPERIMENT, SESSION_GROUPS, download_data.OutputFormat.LATEX
+        )
+        self.assertEqual("application/x-latex", mime_type)
+        self.assertEqual(EXPECTED_LATEX, body)
+
+    def test_json(self):
+        body, mime_type = self._run_handler(
+            EXPERIMENT, SESSION_GROUPS, download_data.OutputFormat.JSON
+        )
+        self.assertEqual("application/json", mime_type)
+        expected_result = {
+            "header": [
+                "initial_temp",
+                "final_temp",
+                "string_hparam",
+                "bool_hparam",
+                "optional_string_hparam",
+                "current_temp",
+                "delta_temp",
+                "optional_metric",
+            ],
+            "rows": [
+                [270.0, 150.0, "a string", True, "", 10.0, 15.0, 33.0],
+                [280.0, 100.0, "AAAAA", False, "", 51.0, 44.5],
```

Comment:
Thanks for adding these test cases! They’re clear and helpful.

It strikes me as a bit odd that optional metrics are silently omitted
from the output rather than having `null` entries or the like. If I had
two or more optional metrics, or optional metrics not at the end of the
list of metrics, how would I know which of the metrics were present?

For instance, if I had hparams `["x"]` and metrics `["y1", "y2", "y3"]`
all optional, how would we represent a row that had only `y1: 1.0` set
vs. a row that had only `y2: 1.0` set? (Same question goes for the other
output formats, too.)

Reply:
Good point, this slipped my mind. I updated the code to explicitly manage this cases.

Reply:
I don’t see any new commits in the PR—forgot to push?